pkg/ottl/grammar.go

```diff
@@ -464,3 +492,43 @@ func buildLexer() *lexer.StatefulDefinition {
 		{Name: "whitespace", Pattern: `\s+`},
 	})
 }
+
+// grammarVisitor allows accessing the grammar AST nodes using the visitor pattern.
+type grammarVisitor interface {
```

Comment:
@edmocosta as I'm reading this PR I feel like I'm right on the edge of full comprehension, but I think I'm getting tripped up by the combining of `checkForCustomError` and checking all the places we visit.  

I know it'll mean traversing the parsed structs twice, but can you separate the new feature from the error checking?

Separately, I like how you've updated the error checking to combine the errors instead of returning the first one found, that is a nice quality of life change and is worthy of its own PR.

Reply:
Sure thing, I will move it to a separate PR.

Reply:
Done, I'll be creating the `checkForCustomError` PR as soon as we merge this one. Thanks!

Reply:
I like this better thank you